query/src/main/java/io/keen/client/java/KeenQueryClient.java

```diff
@@ -0,0 +1,704 @@
+                                   final Map<String, ?> requestData) throws IOException {
+
+        // Build an output source which simply writes the serialized JSON to the output.
+        OutputSource source = new OutputSource() {
+            @Override
+            public void writeTo(OutputStream out) throws IOException {
+                OutputStreamWriter writer = new OutputStreamWriter(out, ENCODING);
+
+                // in queries, requestData may be null.
+                if (requestData != null && requestData.size() != 0) {
+                    jsonHandler.writeJson(writer, requestData);
+                }
+            }
+        };
+
+        // If logging is enabled, log the request being sent.
+        if (KeenLogging.isLoggingEnabled()) {
+            try {
+                StringWriter writer = new StringWriter();
+                jsonHandler.writeJson(writer, requestData);
+                String request = writer.toString();
+                KeenLogging.log(String.format(Locale.US, "Sent request '%s' to URL '%s'",
+                        request, url.toString()));
+            } catch (IOException e) {
+                KeenLogging.log("Couldn't log event written to file: ");
+                e.printStackTrace();
+            }
+        }
+
+        // Send the request.
+        String readkey = project.getReadKey();
+        Request request = new Request(url, "POST", readkey, source, null);
+        Response response = httpHandler.execute(request);
+
+        if (response.isSuccess() == false) {
+            throw new ServerException(response.body);
+        }
+
+        // Parse the response into a map.
+        StringReader reader = new StringReader(response.body);
+        Map<String, Object> responseMap;
+        responseMap = this.jsonHandler.readJson(reader);
+
+        // Get the result object.
+        Object result = responseMap.get(KeenQueryConstants.RESULT);
+        // for successful query, we should get a Result object. But just in case we don't...
+        if (result == null) {
+            String errorCode = responseMap.get(KeenQueryConstants.ERROR_CODE).toString();
+            String message = responseMap.get(KeenQueryConstants.MESSAGE).toString();
+
+            String errorMessage = "Error response received from server";
+            if (errorCode != null) {errorMessage += " " + errorCode;}
+            if (message != null) {errorMessage += ": " + message;}
+
+            throw new KeenQueryClientException(errorMessage);
+        }
+
+
+        return result;
+    }
+
+
+    private String formatBaseURL(String queryName) {
+        return String.format(Locale.US, "%s/%s/projects/%s/queries/%s",
+                baseUrl,
+                KeenConstants.API_VERSION,
+                project.getProjectId(),
+                queryName        // query name
+        );
+    }
+
+    private Long queryResultToLong(QueryResult result) throws KeenQueryClientException {
```

Comment:
The issue here is Double vs. Long. Queries like average, for example, may return a Long value, even though KeenQueryClient's average() method returns a Double (because it can be a Double). So basically the queryResultToDouble() and queryResultToLong() methods just do a little more processing. Perhaps the better solution is that we should restructure our DoubleResult and LongResult to inherit from perhaps a NumericResult class, but the way I did it was just put the extra logic into KeenQueryClient.